### DISCUSSION TITLE
fh.Write_all(wbuf.as_raw())

### DISCUSSION THREAD

**Discussion by aph9944** [mpi4py/mpi4py#412] [Q&A] [answered]
Hello,
I implemented in python a parallel IO with mpi4py. I saw in your code that you use systematically .as_raw() in the write functions such as Write_all, e.g. fh.Write_all(wbuf.as_raw()). Could you please tell me the purpose of .as_raw()? Also when I tried to apply it to my local 3d array, I got the error: 'numpy.ndarray' object has no attribute 'as_raw'
Do you know what the issue might be?
Thanks,
Anthony

**Comment by dalcinl** [comment, marked as answer]
> I saw in your code that you use systematically .as_raw() in the write functions such as Write_all, e.g. fh.Write_all(wbuf.as_raw()). Could you please tell me the purpose of .as_raw()?

Ignore all that. These weird methods you see are all related to testing for the different MPI-related ways to specify buffers. The `as_raw()` methods are defined in custom helper classes for testing that contains an internal array of various types (see `test/arrayimpl.py`). In the particular case of `as_raw()` the internal array is returned plain and raw. There are other methods that help construct buffers specifications with explicit counts and MPI datatypes like `[array, count, datatype]`.

In short: In your own code, just write `fh.Write_all(wbuf)`.

**Reply by aph9944** [reply]
Ok, great. Thanks for your help
